File: apps/wiki/src/rag/classifier.ts
import path from "node:path"

import { Language, programmingExtensions } from "@/rag/languages"

export type ContentKind = "code" | "text" | "csv" | "markdown"

export class ClassifiedFile {
	public kind: ContentKind
	public language?: Language
	public extension: string

	constructor(public path: string) {
		const extension = ClassifiedFile.getFileExtension(path)

		this.extension = extension
		this.kind = ClassifiedFile.extensionToContentKind(extension)
		this.language = programmingExtensions.get(extension)
	}

	static extensionToContentKind(extension: string): ContentKind {
		if (programmingExtensions.has(extension)) return "code"
		if (extension === "csv") return "csv"
		if (extension === "md") return "markdown"

		return "text"
	}

	static getFileExtension(filePath: string): string {
		const ext = path.extname(filePath).toLowerCase()
		return ext ? ext.slice(1) : "empty"
	}
}

File: apps/wiki-rs/src/rag/classifier.rs
use std::path::Path;
use uuid::Uuid;

#[derive(Clone)]
pub struct ClassifiedFile {
    pub project_id: Uuid,
    pub relative_path: String,
    pub kind: String,
    pub language: String,
}

pub fn classify_file(project_id: Uuid, root: &Path, file_path: &Path) -> Option<ClassifiedFile> {
    let relative = file_path.strip_prefix(root).ok()?;
    let relative_path = relative.to_string_lossy().replace('\\', "/");

    if relative_path.contains("/.git/") || relative_path.ends_with("/.git") {
        return None;
    }

    let extension = file_path
        .extension()
        .and_then(|value| value.to_str())
        .unwrap_or_default()
        .to_ascii_lowercase();

    let (kind, language) = classify_extension(&extension);

    Some(ClassifiedFile {
        project_id,
        relative_path,
        kind: kind.to_string(),
        language: language.to_string(),
    })
}

fn classify_extension(extension: &str) -> (&'static str, &'static str) {
    match extension {
        "ts" => ("code", "typescript"),
        "tsx" => ("code", "typescript-react"),
        "js" => ("code", "javascript"),
        "jsx" => ("code", "javascript-react"),
        "py" => ("code", "python"),
        "go" => ("code", "go"),
        "java" => ("code", "java"),
        "c" | "h" => ("code", "c"),
        "cpp" | "hpp" => ("code", "cpp"),
        "rs" => ("code", "rust"),
        "proto" => ("code", "proto"),
        "md" => ("markdown", "none"),
        "csv" => ("csv", "none"),
        _ => ("text", "none"),
    }
}
